core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -467,6 +502,26 @@ class Partition(val topic: String,
     info
   }
 
+  /**
+   * Update logStartOffset and low watermark if 1) purgeOffset <= highWatermark and 2) it is the leader replica.
+   * This function can trigger log segment deletion and log rolling.
+   *
+   * Return low watermark of the partition.
+   */
+  def purgeRecordsOnLeader(offset: Long): Long = {
+    inReadLock(leaderIsrUpdateLock) {
```

Comment:
I think we should use readlock since this method doesn't update leader or isr of the partition, right?